pkg/sql/information_schema.go

```diff
@@ -1529,6 +1529,198 @@ const (
+	ctx context.Context,
+	p *planner,
+	dbContext *DatabaseDescriptor,
+	virtualOpts virtualOpts,
+	gen tableIterationGenerator,
+) (virtualTableGenerator, error) {
+	return forEachTableDescAllGeneratorWithLookupInternal(
+		ctx,
+		p,
+		dbContext,
+		virtualOpts,
+		true, /* allowAdding */
+		func(
+			db *sqlbase.DatabaseDescriptor,
+			scName string,
+			table *sqlbase.TableDescriptor,
+			_ tableLookupFn,
+			iter int,
+		) (tree.Datums, error) {
+			return gen(db, scName, table, iter)
+		},
+	)
+}
+
+// To get rid of lint messages.
+var _ = forEachTableDescAllGeneratorWithLookup
+
+// forEachTableDescAllGenerator takes in a tableIterationGeneratorWithLookup
+// and returns a virtualTableGenerator that applies the tableIterationGenerator
+// to all tables.
+func forEachTableDescAllGeneratorWithLookup(
+	ctx context.Context,
+	p *planner,
+	db *DatabaseDescriptor,
+	virtualOpts virtualOpts,
+	gen tableIterationGeneratorWithLookup,
+) (virtualTableGenerator, error) {
+	return forEachTableDescAllGeneratorWithLookupInternal(
+		ctx,
+		p,
+		db,
+		virtualOpts,
+		true, /* allowAdding */
+		gen,
+	)
+}
+
+// forEachTableDescAllGeneratorWithLookupInternal takes in a
+// tableIterationGeneratorWithLookup and returns a virtualTableGenerator
+// that lazily applies the tableIterationGeneratorWithLookup to all tables.
+func forEachTableDescAllGeneratorWithLookupInternal(
+	ctx context.Context,
+	p *planner,
+	dbContext *DatabaseDescriptor,
+	virtualOpts virtualOpts,
+	allowAdding bool,
+	gen tableIterationGeneratorWithLookup,
+) (virtualTableGenerator, error) {
+	// Eagerly lookup data we need, like physical and virtual
+	// table descriptors.
+	descs, err := p.Tables().getAllDescriptors(ctx, p.Txn())
+	if err != nil {
+		return nil, err
+	}
+	lCtx := newInternalLookupCtx(descs, dbContext)
+	var vEntries map[string]virtualSchemaEntry
+	var vSchemaNames []string
+	if virtualOpts == virtualMany || virtualOpts == virtualOnce {
+		vt := p.getVirtualTabler()
+		vEntries = vt.getEntries()
+		vSchemaNames = vt.getSchemaNames()
+	}
+	schemaNames, err := getSchemaNames(ctx, p, dbContext)
+	if err != nil {
+		return nil, err
+	}
+
+	// Indexes into different streams to remember where we are in iteration.
+	virtualDBIdx := 0
+	virtualSchemaIdx := 0
+	virtualTableIdx := 0
+	physTableIdx := 0
+
+	// iterCount holds how many times we have processed this same set
+	// of (db, schema, table).
+	iterCount := 0
+
+	// iterateVirtual perfoms a lazy iteration over the set of virtual tables.
+	iterateVirtual := func(db *DatabaseDescriptor) (tree.Datums, error) {
+		for virtualSchemaIdx < len(vSchemaNames) {
+			vSchemaName := vSchemaNames[virtualSchemaIdx]
+			e := vEntries[vSchemaName]
+			for virtualTableIdx < len(e.orderedDefNames) {
+				te := e.defs[e.orderedDefNames[virtualTableIdx]]
+				res, err := gen(nil, vSchemaName, te.desc, lCtx, iterCount)
+				if err != nil {
+					return nil, err
+				}
+				if res == nil {
+					virtualTableIdx++
+					iterCount = 0
```

Comment:
Yuck... could you try using a channel instead? I think this would be a good use case for a goroutine that pushes into an unbuffered channel.

Reply:
Is it really that ugly? It feels just kind of the nature for how these things are. 

Re the channel idea:
* We'd have to manage cleanup for this goroutine -- what do we do if we only read one row from the virtual table? 
* Having to sync across the channel for each row seems pretty slow...

Reply:
Passing stuff between channels I think is not so slow, and it's cleaner than all of this index accounting which I feel is very error prone. There are several patterns available for cleaning up channels, but you can do it with a defer without too much trouble, right?

Reply:
Let me know if i have an unnecessary worry:

Do we always scan all the rows in a virtual table? If not, say we start this gorountine and then only read one row from our generator (and then never call it again). How does the generator know to signal to the gorountine that it is done? Or is this something that could just be easily returned for the virtual table scanner to run on close.

Reply:
A common pattern is to pass in a cancellation channel to the goroutine that's generating stuff, and then do a select on both channels, or to use a context.WithCancel.

So you'd have something like:

```
for {
// fetch chunk of rows
for row in chunk {
select {
case <-ctx.Done():
   return // also clean up any remaining state from within the generator goroutine
case outCh <- theRow:
   // do nothing, continue in the loop
}
}
}
```